File: kgra/src/agent/state.py
"""
KGRA State Definition — TypedDict for the LangGraph state machine.
Matches Section 5 of the profile, extended with evaluation state.
"""

from __future__ import annotations
from typing import TypedDict, Optional, Any
from enum import Enum


class IntentType(str, Enum):
    FACTUAL = "factual"
    COMPARATIVE = "comparative"
    EXPLORATORY = "exploratory"
    TEMPORAL = "temporal"
    HYPOTHETICAL = "hypothetical"
    LEARNING_REQUEST = "learning_request"
    RESEARCH_DIRECTIVE = "research_directive"
    ONTOLOGY_DESIGN = "ontology_design"
    CONTRADICTION_RESOLUTION = "contradiction_resolution"
    CROSS_AGENT = "cross_agent_collaboration"
    REASONING_PATH_REUSE = "reasoning_path_reuse"
    SELF_EVALUATION = "self_evaluation"
    BUNDLE_EVALUATION = "bundle_evaluation"


class RuntimeMode(str, Enum):
    DIRECT_QUERY = "direct_query"
    PATH_REASONING = "path_reasoning"
    GRAPHRAG_LOCAL = "graphrag_local"
    GRAPHRAG_GLOBAL = "graphrag_global"
    DRIFT = "drift"
    BASIC_RAG = "basic_rag"
    HUMAN_REVIEW = "human_review"
    SELF_LEARNING = "self_learning"
    EXPERTISE_BUILDING = "expertise_building"
    RESEARCH_MODE = "research_mode"
    CROSS_AGENT = "cross_agent_collaboration"
    REASONING_PATH_REUSE = "reasoning_path_reuse"
    BUNDLE_EVALUATION = "bundle_evaluation"


class EvidenceNode(TypedDict, total=False):
    id: str
    node_type: str  # "fact" | "claim" | "evidence" | "decision"
    label: str
    source_ref: str
    confidence: float
    timestamp: str
    properties: dict


class EvidenceEdge(TypedDict, total=False):
    id: str
    from_id: str
    to_id: str
    edge_type: str  # "supports" | "contradicts" | "resolves" | "depends_on" | "precedes"
    weight: float


class ProvenanceEntry(TypedDict, total=False):
    source_type: str  # "graph" | "document" | "memory" | "other_ai" | "research" | "evaluated_bundle"
    source_ref: str
    operation_ref: str
    derivation_path: list[str]


class CostTracker(TypedDict, total=False):
    llm_tokens: int
    graph_reads: int
    vector_queries: int
    mcp_calls: int
    estimated_usd_cents: float


class TemporalScope(TypedDict, total=False):
    as_of: str
    valid_from: str
    valid_to: str
    warning: str


class GovernanceVerdict(TypedDict, total=False):
    verdict: str  # "allow" | "conditions" | "deny"
    reasons: list[str]
    human_review_required: bool


class BundleEvaluationState(TypedDict, total=False):
    bundle_id: str
    documents: list[dict]
    review_log: str
    coherence_score: float
    coherence_verdict: str
    readiness_score: float
    readiness_missing: list[str]
    resolved_decisions: list[dict]
    contract_mappings: list[dict]
    current_pass: int  # 1=gist, 2=structure, 3=gaps


class LearningOutcome(TypedDict, total=False):
    knowledge_graph_updated: bool
    new_node_types: list[str]
    new_relation_types: list[str]
    reasoning_path_id: str
    other_ai_contributions: list[dict]
    evaluated_bundle: dict


class AnalyticalMeta(TypedDict, total=False):
    reasoning_strength: float
    vulnerability: float
    alternative_hypotheses: list[dict]
    causal_claims: list[dict]


class KGRAState(TypedDict, total=False):
    """Complete state for the KGRA LangGraph state machine."""

    # ── Request ──
    run_id: str
    request_id: str
    query: str
    workspace_id: str
    session_id: str
    user_id: str

    # ── Intent & Mode ──
    intent: str  # IntentType value
    resolved_terms: dict[str, str]  # domain term → canonical form
    mode: str  # RuntimeMode value

    # ── Schema & Query Planning ──
    schema_slice: dict
    planned_query: str
    query_language: str  # "cypher" | "sparql" | "graphrag" | "vector"
    query_validated: bool
    validation_errors: list[str]

    # ── Execution Results ──
    raw_results: list[dict]
    paths: list[dict]
    ranked_paths: list[dict]

    # ── Evidence ──
    observed_facts: list[dict]
    inferred_claims: list[dict]
    evidence_nodes: list[EvidenceNode]
    evidence_edges: list[EvidenceEdge]
    provenance: list[ProvenanceEntry]

    # ── Temporal ──
    temporal: TemporalScope

    # ── Contradictions ──
    contradictions: list[dict]
    hypotheses: list[dict]

    # ── Answer ──
    answer: str
    confidence: float
    missing_knowledge: dict

    # ── Governance ──
    governance: GovernanceVerdict

    # ── Learning ──
    learning: LearningOutcome

    # ── Analytical ──
    analytical: AnalyticalMeta

    # ── Bundle Evaluation (subgraph state) ──
    evaluation: BundleEvaluationState

    # ── Cost ──
    cost: CostTracker

    # ── Control Flow ──
    current_node: str
    completed_nodes: list[str]
    errors: list[str]
    tool_calls: list[dict]  # log of every tool invocation
    started_at: str
    fallback_used: str

File: server/kgra-agent/state.ts
/**
 * KGRA State — TypeScript port of the Python KGRAState.
 * Complete state for the KGRA state machine pipeline.
 */

// ── Intent Types ────────────────────────────────────────────────

export type IntentType =
  | "factual"
  | "comparative"
  | "exploratory"
  | "temporal"
  | "hypothetical"
  | "learning_request"
  | "research_directive"
  | "ontology_design"
  | "contradiction_resolution"
  | "cross_agent_collaboration"
  | "reasoning_path_reuse"
  | "self_evaluation"
  | "bundle_evaluation";

// ── Runtime Modes ───────────────────────────────────────────────

export type RuntimeMode =
  | "direct_query"
  | "path_reasoning"
  | "graphrag_local"
  | "graphrag_global"
  | "drift"
  | "basic_rag"
  | "human_review"
  | "self_learning"
  | "expertise_building"
  | "research_mode"
  | "cross_agent_collaboration"
  | "reasoning_path_reuse"
  | "bundle_evaluation";

// ── Evidence ────────────────────────────────────────────────────

export interface EvidenceNode {
  id: string;
  node_type: string; // "fact" | "claim" | "evidence" | "decision"
  label: string;
  source_ref: string;
  confidence: number;
  timestamp?: string;
  properties?: Record<string, unknown>;
}

export interface EvidenceEdge {
  id: string;
  from_id: string;
  to_id: string;
  edge_type: string; // "supports" | "contradicts" | "resolves" | "depends_on" | "precedes"
  weight: number;
}

// ── Provenance ──────────────────────────────────────────────────

export interface ProvenanceEntry {
  source_type: string; // "graph" | "document" | "memory" | "llm" | "research" | "evaluated_bundle"
  source_ref: string;
  operation_ref: string;
  derivation_path?: string[];
}

// ── Cost ────────────────────────────────────────────────────────

export interface CostTracker {
  llm_tokens: number;
  graph_reads: number;
  vector_queries: number;
  mcp_calls: number;
  estimated_usd_cents: number;
}

// ── Temporal ────────────────────────────────────────────────────

export interface TemporalScope {
  as_of: string;
  valid_from?: string;
  valid_to?: string;
  warning?: string;
}

// ── Governance ──────────────────────────────────────────────────

export interface GovernanceVerdict {
  verdict: string; // "allow" | "conditions" | "deny"
  reasons: string[];
  human_review_required: boolean;
}

// ── Bundle Evaluation ───────────────────────────────────────────

export interface BundleEvaluationState {
  bundle_id: string;
  documents: Array<{ name: string; content: string }>;
  review_log: string;
  coherence_score: number;
  coherence_verdict: string;
  readiness_score: number;
  readiness_missing: string[];
  resolved_decisions: Array<Record<string, unknown>>;
  contract_mappings: Array<Record<string, unknown>>;
  current_pass: number;
}

// ── Learning ────────────────────────────────────────────────────

export interface LearningOutcome {
  knowledge_graph_updated: boolean;
  new_node_types?: string[];
  new_relation_types?: string[];
  reasoning_path_id?: string;
  evaluated_bundle?: Record<string, unknown>;
}

// ── Analytical ──────────────────────────────────────────────────

export interface AnalyticalMeta {
  reasoning_strength: number;
  vulnerability?: number;
  alternative_hypotheses?: Array<Record<string, unknown>>;
  causal_claims?: Array<Record<string, unknown>>;
}

// ── Tool Call Log ───────────────────────────────────────────────

export interface ToolCallEntry {
  tool: string;
  inputs: Record<string, unknown>;
  output_summary: string;
  timestamp: string;
  cost: Partial<CostTracker>;
}

// ── Complete KGRA State ─────────────────────────────────────────

export interface KGRAState {
  // Request
  run_id: string;
  request_id: string;
  query: string;
  workspace_id?: string;
  session_id?: string;
  user_id?: string;

  // Intent & Mode
  intent: IntentType;
  resolved_terms: Record<string, string>;
  mode: RuntimeMode;

  // Schema & Query Planning
  schema_slice: Record<string, unknown>;
  planned_query: string;
  query_language: string;
  query_validated: boolean;
  validation_errors: string[];

  // Execution Results
  raw_results: Array<Record<string, unknown>>;
  paths: Array<Record<string, unknown>>;
  ranked_paths: Array<Record<string, unknown>>;

  // Evidence
  observed_facts: Array<{ id: string; label: string; sourceRef: string }>;
  inferred_claims: Array<{ id: string; text: string; basis: string[]; confidence: number }>;
  evidence_nodes: EvidenceNode[];
  evidence_edges: EvidenceEdge[];
  provenance: ProvenanceEntry[];

  // Temporal
  temporal: TemporalScope;

  // Contradictions
  contradictions: Array<Record<string, unknown>>;
  hypotheses: Array<Record<string, unknown>>;

  // Answer
  answer: string;
  confidence: number;
  missing_knowledge: { hasGap: boolean; summary?: string };

  // Governance
  governance: GovernanceVerdict;

  // Learning
  learning: LearningOutcome;

  // Analytical
  analytical: AnalyticalMeta;

  // Bundle Evaluation
  evaluation?: BundleEvaluationState;

  // Cost
  cost: CostTracker;

  // Control Flow
  current_node: string;
  completed_nodes: string[];
  errors: string[];
  tool_calls: ToolCallEntry[];
  started_at: string;
  fallback_used: string;
}

// ── Factory ─────────────────────────────────────────────────────

export function createInitialState(query: string, runId: string, requestId: string): KGRAState {
  const now = new Date().toISOString();
  return {
    run_id: runId,
    request_id: requestId,
    query,
    intent: "factual",
    resolved_terms: {},
    mode: "direct_query",
    schema_slice: {},
    planned_query: "",
    query_language: "",
    query_validated: false,
    validation_errors: [],
    raw_results: [],
    paths: [],
    ranked_paths: [],
    observed_facts: [],
    inferred_claims: [],
    evidence_nodes: [],
    evidence_edges: [],
    provenance: [],
    temporal: { as_of: now },
    contradictions: [],
    hypotheses: [],
    answer: "",
    confidence: 0,
    missing_knowledge: { hasGap: false },
    governance: { verdict: "allow", reasons: [], human_review_required: false },
    learning: { knowledge_graph_updated: false },
    analytical: { reasoning_strength: 0 },
    cost: { llm_tokens: 0, graph_reads: 0, vector_queries: 0, mcp_calls: 0, estimated_usd_cents: 0 },
    current_node: "",
    completed_nodes: [],
    errors: [],
    tool_calls: [],
    started_at: now,
    fallback_used: "",
  };
}
